vng_api_common/caching/models.py

```diff
@@ -0,0 +1,25 @@
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+
+from .etags import calculate_etag
+
+
+class ETagMixin(models.Model):
+    """
+    Automatically calculate the (new) ETag value on save.
+    """
+
+    _etag = models.CharField(
+        _("etag value"),
+        max_length=32,
+        help_text=_("MD5 hash of the resource representation in its current version."),
+        editable=False,
+    )
+
+    class Meta:
+        abstract = True
+
+    def save(self, *args, **kwargs):
+        # TODO: change into post-save to handle pk?
+        self._etag = calculate_etag(self)
```

Comment:
Je rendered hierdoor wel een unsaved instance. Kan dat niet tot problemen lijden als er bijv. relaties zijn om te renderen en er is nog geen PK? m2m vooral.

Reply:
Ja, deze vind ik ook een beetje naar nog. Ik had een poging gedaan om te forceren dat er een PK was, maar dat ging niet lekker. Denk dat dit moet verhuizen naar een `post_save` signal en een `@transaction.atomic` op de `def save` moet komen :thinking:

Reply:
Gaat voor ons overigens nu wel goed omdat `uuid` wel gezet is en de `pk` niet in de serializer output voorkomt.

Met M2M gaat het stuk gaan, maar daar loop ik dan wel weer vanzelf tegenaan als ik aan catalogus API kom... :thinking:

Reply:
Sorry voor de spam...

Waarschijnlijk best om dit hele verhaal in te laten haken op `transaction.on_commit` - op die manier zijn we zeker dat alle relaties gecommit zijn, en we het "eindresultaat" hebben. En dan vermijden we ook signals.

Reply:
Ideeen:

- Zet etag op None bij elke save en bereken etag als deze None is bij de eerste GET
- Pre-save en Post-save signals. Signals vermijden is voor mij geen doel op zich EN je lost het generiek op zonder elke save method te hoeven overriden. Pre-save als er al een PK is, Post-save als die er nog niet is.
- Is het niet vreemd dat je dit op model niveau wil regelen terwijl het helemaal gaat over een API-response? Ik denk dat dit gewoon niet de goede plek is om het te doen, maar meer in een post_save-hook van de view of serializer. In het begin wilde we ook notificaties op model-niveau doen, maar dat is gewoon niet de goede plek en hebben het verhuist naar de api laag (wat correct is).

Reply:
bij eerste GET berekenen is volgens mij enige correct oplossing. `save`, `pre_save` en `post_save` zijn eigenlijk allemaal in hetzelfde bedje ziek: je gaat `bulk_create` etc. missen.

Dus in een `GET` of `HEAD` moet die berekend worden inderdaad als die er niet is, en bij elke data-change moet die gecleared worden. Wat op zich nog weer uitdagend is met m2m, want als de relatie verandert in een ander model, dan heeft dit effect op de resource die de m2m ontsluit, terwijl die zelf helemaal niet getriggerd wordt. Lastig... :thinking:

Reply:
> Wat op zich nog weer uitdagend is met m2m, want als de relatie verandert in een ander model, dan heeft dit effect op de resource die de m2m ontsluit, terwijl die zelf helemaal niet getriggerd wordt.

Zelfde geldt voor omgekeerde FK relaties.

Lastig maar ook wel weer overzichtelijk: Elke FK en M2M moet eigenlijk gevolgd worden en etags resetten.

**Idee**: Kunnen we niet iets soortgelijks doen als bij notificaties, met het hoofdobject? We geven elke resource de etag van het hoofdobject, en bij elke update van wat dan ook binnen die API, updaten we de etag van het hoofdobject (wat dan automatisch ook bij andere resources geupdate wordt). Zo weten we zeker dat er niet per ongeluk door relaties een etag ongewijzigd blijft.

Ja, dit zorgt er voor dat sommige resources een andere etag krijgen ook al is er niets verandert maar het voorkomt een hoop mogelijke problemen. Daarnaast gaat dit op oudere zaken totaal geen effect hebben en wordt er alleen wat extra's gedaan bij "actievere" zaken.

Reply:
ETags kan je niet limiteren tot "wat er in de API gebeurt". Wijzigingen in background jobs, admin interfaces... die moeten ook allemaal leiden tot een nieuwe ETag value. Notificaties is puur op basis van events die gebeuren in de API.

Ik heb nu iig het omgekat, moet nog een paar testcases toevoegen voor reverse FKs, maar dan is het wel in orde. m2m add/remove/clear werken, in beide richtingen, op een generieke manier.